File: hsp/hsp/strbuf.m
//
//	HSP3 string support
//	(おおらかなメモリ管理をするバッファマネージャー)
//	(sbAllocでSTRBUF_BLOCKSIZEのバッファを確保します)
//	(あとはsbCopy,sbAddで自動的にバッファの再確保を行ないます)
//	onion software/onitama 2004/6
//
#include "strbuf.h"
#include "debug_message.h"
#include "hsp3struct_debug.h"
#include "supio_hsp3.h"
#import <Cocoa/Cocoa.h>
#import <Foundation/Foundation.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define REALLOC realloc
#define MALLOC malloc
#define FREE free
/*------------------------------------------------------------*/
/*
 system data
 */
/*------------------------------------------------------------*/
typedef struct {
    STRBUF* mem;
    int len;
} SLOT;
static SLOT* mem_sb;
static int str_blockcur;
static int slot_len;
static STRBUF* freelist;
// STRINF_FLAG_NONE のとき STRINF::extptr を free list の次のポインタに使う
#define STRINF_NEXT(inf) ((inf).extptr)
#define STRBUF_NEXT(buf) STRINF_NEXT((buf)->inf)
#define GET_INTINF(buf) (&((buf)->inf.intptr->inf))
@implementation ViewController (strbuf)
/*------------------------------------------------------------*/
/*
 internal function
 */
/*------------------------------------------------------------*/
-(void)BlockPtrPrepare {
    STRBUF* sb;
    if (str_blockcur == 0) {
        mem_sb = (SLOT*)MALLOC(sizeof(SLOT));
    } else {
        mem_sb = (SLOT*)REALLOC(mem_sb, sizeof(SLOT) * (str_blockcur + 1));
    }
    sb = (STRBUF*)MALLOC(sizeof(STRBUF) * slot_len);
    if (sb == NULL) {
        NSString* error_str =
        [NSString stringWithFormat:@"%d", HSPERR_OUT_OF_MEMORY];
        @throw [NSException exceptionWithName:@"" reason:error_str userInfo:nil];
    }
    STRBUF* p = sb;
    STRBUF* pend = p + slot_len;
    mem_sb[str_blockcur].mem = sb;
    mem_sb[str_blockcur].len = slot_len;
    str_blockcur++;
    slot_len = (int)(slot_len * 1.8);
    while (p < pend) {
        p->inf.intptr = p;
        p->inf.flag = STRINF_FLAG_NONE;
        STRBUF_NEXT(p) = freelist;
        freelist = p;
        p++;
    }
}
-(STRBUF*)BlockEntry {
    //		空きエントリーブロックを探す
    //
    if (freelist == NULL) {
        [self BlockPtrPrepare];
    }
    STRBUF* buf = freelist;
    freelist = STRBUF_NEXT(freelist);
    return buf;
}
-(char*)BlockAlloc:(int)size {
    int* p;
    STRBUF* st;
    STRBUF* st2;
    STRINF* inf;
    st = [self BlockEntry];
    inf = &(st->inf);
    if (size <= STRBUF_BLOCKSIZE) {
        inf->flag = STRINF_FLAG_USEINT;
        inf->size = STRBUF_BLOCKSIZE;
        p = (int*)st->data;
        inf->ptr = (char*)p;
    } else {
        inf->flag = STRINF_FLAG_USEEXT;
        inf->size = size;
        st2 = (STRBUF*)MALLOC(size + sizeof(STRINF));
        p = (int*)(st2->data);
        inf->extptr = st2;
        inf->ptr = (char*)p;
        st2->inf = *inf;
    }
    *p = 0;
    // return inf->ptr;
    return (char*)p;
}
-(void)FreeExtPtr:(STRINF*)inf {
    if (inf->flag == STRINF_FLAG_USEEXT) {
        FREE(inf->extptr);
    }
}
-(void)BlockFree:(STRINF*)inf {
    [self FreeExtPtr:inf];
    STRINF_NEXT(*inf) = freelist;
    freelist = (STRBUF*)inf;
    inf->flag = STRINF_FLAG_NONE;
}
-(char*)BlockRealloc:(STRBUF*)st size:(int)size {
    char* p;
    STRINF* inf;
    STRBUF* newst;
    inf = GET_INTINF(st);
    if (size <= inf->size)
        return inf->ptr;
    newst = (STRBUF*)MALLOC(size + sizeof(STRINF));
    p = newst->data;
    memcpy(p, inf->ptr, inf->size);
    [self FreeExtPtr:inf];
    inf->size = size;
    inf->flag = STRINF_FLAG_USEEXT;
    inf->ptr = p;
    inf->extptr = newst;
    newst->inf = *inf;
    return p;
}
-(void)BlockInfo:(STRINF*)inf {
    STRBUF* newst;
    if (inf->flag == STRINF_FLAG_USEEXT) {
        newst = (STRBUF*)inf->extptr;
    }
}
/*------------------------------------------------------------*/
/*
 interface
 */
/*------------------------------------------------------------*/
-(void)sbInit {
    str_blockcur = 0;
    freelist = NULL;
    slot_len = STRBUF_BLOCK_DEFAULT;
    [self BlockPtrPrepare];
}
-(void)sbBye {
    int i;
    for (i = 0; i < str_blockcur; i++) {
        STRBUF* mem = mem_sb[i].mem;
        STRBUF* p = mem;
        STRBUF* pend = p + mem_sb[i].len;
        while (p < pend) {
            [self FreeExtPtr:&p->inf];
            p++;
        }
        FREE(mem);
    }
    FREE(mem_sb);
}
-(STRINF*)sbGetSTRINF:(char*)ptr {
    return (STRINF*)(ptr - sizeof(STRINF));
}
-(char*)sbAlloc:(int)size {
    int sz;
    sz = size;
    if (size < STRBUF_BLOCKSIZE)
        sz = STRBUF_BLOCKSIZE;
    return [self BlockAlloc:sz];
}
-(char*)sbAllocClear:(int)size {
    char* p;
    p = [self sbAlloc:size];
    memset(p, 0, size);
    return p;
}
-(void)sbFree:(void*)ptr {
    char* p;
    STRBUF* st;
    STRINF* inf;
    p = (char*)ptr;
    st = (STRBUF*)(p - sizeof(STRINF));
    inf = GET_INTINF(st);
    if (p != (inf->ptr)) {
        return;
    }
    [self BlockFree:inf];
}
-(char*)sbExpand:(char*)ptr size:(int)size {
    STRBUF* st;
    st = (STRBUF*)(ptr - sizeof(STRINF));
    return [self BlockRealloc:st size:size];
}
-(void)sbCopy:(char**)pptr data:(char*)data size:(int)size {
    int sz;
    char* ptr;
    char* p;
    STRBUF* st;
    ptr = *pptr;
    st = (STRBUF*)(ptr - sizeof(STRINF));
    sz = st->inf.size;
    p = st->inf.ptr;
    if (size > sz) {
        p = [self BlockRealloc:st size:size];
        *pptr = p;
    }
    memcpy(p, data, size);
}
-(void)sbAdd:(char**)pptr data:(char*)data size:(int)size mode:(int)mode {
    //		mode:0=normal/1=string
    int sz, newsize;
    STRBUF* st;
    char* ptr;
    char* p;
    ptr = *pptr;
    st = (STRBUF*)(ptr - sizeof(STRINF));
    p = st->inf.ptr;
    if (mode) {
        sz = (int)strlen(p); // 文字列データ
    } else {
        sz = st->inf.size; // 通常データ
    }
    newsize = sz + size;
    if (newsize > (st->inf.size)) {
        newsize = (newsize + 0xfff) & 0xfffff000; // 8K単位で確保
        // Alertf( "#Alloc%d",newsize );
        p = [self BlockRealloc:st size:newsize];
        *pptr = p;
    }
    memcpy(p + sz, data, size);
}
-(void)sbStrCopy:(char**)ptr str:(char*)str {
    [self sbCopy:ptr data:str size:(int)strlen(str) + 1];
}
-(void)sbStrAdd:(char**)ptr str:(char*)str {
    [self sbAdd:ptr data:str size:(int)strlen(str) + 1 mode:1];
}
-(void*)sbGetOption:(char*)ptr {
    STRBUF* st;
    st = (STRBUF*)(ptr - sizeof(STRINF));
    return st->inf.opt;
}
-(void)sbSetOption:(char*)ptr option:(void*)option {
    STRBUF* st;
    STRINF* inf;
    st = (STRBUF*)(ptr - sizeof(STRINF));
    st->inf.opt = option;
    inf = GET_INTINF(st);
    inf->opt = option;
}
@end
/*
 void sbInfo( char *ptr )
 {
 STRBUF *st;
 st = (STRBUF *)( ptr - sizeof(STRINF) );
 Alertf( "size:%d (%x)",st->inf.size, st->inf.ptr );
 }
 */
//@end

File: hsp-engine/hsp/strbuf.c
//
//	文字列バッファマネージャー
//

#include "strbuf.h"

static strbuf_memory_t *mem_sb;
static int str_blockcur;
static int slot_len;
static strbuf_t *freelist;

// STRINF_FLAG_NONE のとき STRINF::extptr を free list の次のポインタに使う
#define STRINF_NEXT(inf) ((inf).extptr)
#define STRBUF_NEXT(buf) STRINF_NEXT((buf)->inf)
#define GET_INTINF(buf) (&((buf)->inf.intptr->inf))

//------------------------------------------------------------
// internal function
//------------------------------------------------------------

static void prepare_block_ptr(void) {
    if (str_blockcur == 0) {
        mem_sb = (strbuf_memory_t *) malloc(sizeof(strbuf_memory_t));
    } else {
        mem_sb = (strbuf_memory_t *) realloc(mem_sb, sizeof(strbuf_memory_t) * (str_blockcur + 1));
    }

    strbuf_t *sb = (strbuf_t *) malloc(sizeof(strbuf_t) * slot_len);
    if (sb == NULL) {
        fprintf(stderr, "Error: %d\n", HSPERR_OUT_OF_MEMORY);
        exit(EXIT_FAILURE);
    }

    strbuf_t *p = sb;
    strbuf_t *pend = p + slot_len;

    mem_sb[str_blockcur].mem = sb;
    mem_sb[str_blockcur].len = slot_len;
    str_blockcur++;
    slot_len = (int) (slot_len * 1.8);

    while (p < pend) {
        p->inf.intptr = p;
        p->inf.flag = STRINF_FLAG_NONE;
        STRBUF_NEXT(p) = freelist;
        freelist = p;
        p++;
    }
}

// 空きエントリーブロックを探す
static strbuf_t *search_entry_block(void) {
    if (freelist == NULL) {
        prepare_block_ptr();
    }
    strbuf_t *buf = freelist;
    freelist = STRBUF_NEXT(freelist);
    return buf;
}

static char *alloc_block(int size) {
    int *p;
    strbuf_t *st = search_entry_block();
    strbuf_t *st2;
    strbuf_info_t *inf = &(st->inf);
    if (size <= STRBUF_BLOCKSIZE) {
        inf->flag = STRINF_FLAG_USEINT;
        inf->size = STRBUF_BLOCKSIZE;
        p = (int *) st->data;
        inf->ptr = (char *) p;
    } else {
        inf->flag = STRINF_FLAG_USEEXT;
        inf->size = size;
        st2 = (strbuf_t *) malloc(size + sizeof(strbuf_info_t));
        p = (int *) (st2->data);
        inf->extptr = st2;
        inf->ptr = (char *) p;
        st2->inf = *inf;
    }
    *p = 0;
    return (char *) p;
}

static void free_ext_ptr(strbuf_info_t *inf) {
    if (inf->flag == STRINF_FLAG_USEEXT) {
        free(inf->extptr);
    }
}

static void free_block(strbuf_info_t *inf) {
    free_ext_ptr(inf);
    STRINF_NEXT(*inf) = freelist;
    freelist = (strbuf_t *) inf;
    inf->flag = STRINF_FLAG_NONE;
}

static char *realloc_block(strbuf_t *st, int size) {
    strbuf_info_t *inf = GET_INTINF(st);

    if (size <= inf->size)
        return inf->ptr;
    strbuf_t *newst = (strbuf_t *) malloc(size + sizeof(strbuf_info_t));
    char *p = newst->data;

    memcpy(p, inf->ptr, inf->size);
    free_ext_ptr(inf);
    inf->size = size;
    inf->flag = STRINF_FLAG_USEEXT;
    inf->ptr = p;
    inf->extptr = newst;
    newst->inf = *inf;
    return p;
}

void block_info(strbuf_info_t *inf) {
    strbuf_t *newst;
    if (inf->flag == STRINF_FLAG_USEEXT) {
        newst = (strbuf_t *) inf->extptr;
    }
}

//------------------------------------------------------------
// interface
//------------------------------------------------------------

void strbuf_init(void) {
    str_blockcur = 0;
    freelist = NULL;
    slot_len = STRBUF_BLOCK_DEFAULT;
    prepare_block_ptr();
}

void strbuf_bye(void) {
    for (int i = 0; i < str_blockcur; i++) {
        strbuf_t *mem = mem_sb[i].mem;
        strbuf_t *p = mem;
        strbuf_t *pend = p + mem_sb[i].len;
        while (p < pend) {
            free_ext_ptr(&p->inf);
            p++;
        }
        free(mem);
    }
    free(mem_sb);
}

strbuf_info_t *strbuf_get_strbuf_info(char *ptr) {
    return (strbuf_info_t *) (ptr - sizeof(strbuf_info_t));
}

char *strbuf_alloc(int size) {
    int sz = size;
    if (size < STRBUF_BLOCKSIZE)
        sz = STRBUF_BLOCKSIZE;
    return alloc_block(sz);
}

char *strbuf_alloc_clear(int size) {
    char *p = strbuf_alloc(size);
    memset(p, 0, size);
    return p;
}

void strbuf_free(void *ptr) {
    char *p = (char *) ptr;
    strbuf_t *st = (strbuf_t *) (p - sizeof(strbuf_info_t));
    strbuf_info_t *inf = GET_INTINF(st);
    if (p != (inf->ptr)) {
        return;
    }
    free_block(inf);
}

char *strbuf_expand(char *ptr, int size) {
    strbuf_t *st = (strbuf_t *) (ptr - sizeof(strbuf_info_t));
    return realloc_block(st, size);
}

void strbuf_copy(char **pptr, char *data, int size) {
    char *ptr = *pptr;
    strbuf_t *st = (strbuf_t *) (ptr - sizeof(strbuf_info_t));
    int sz = st->inf.size;
    char *p = st->inf.ptr;
    if (size > sz) {
        p = realloc_block(st, size);
        *pptr = p;
    }
    memcpy(p, data, size);
}

// mode:0=normal/1=string
void strbuf_add(char **pptr, char *data, int size, int mode) {
    char *ptr = *pptr;
    strbuf_t *st = (strbuf_t *) (ptr - sizeof(strbuf_info_t));
    char *p = st->inf.ptr;
    int sz;

    if (mode) {
        sz = (int) strlen(p); // 文字列データ
    } else {
        sz = st->inf.size; // 通常データ
    }

    int newsize = sz + size;
    if (newsize > (st->inf.size)) {
        newsize = (newsize + 0xfff) & 0xfffff000; // 8K単位で確保
        p = realloc_block(st, newsize);
        *pptr = p;
    }
    memcpy(p + sz, data, size);
}

void strbuf_copy_str(char **ptr, char *str) {
    strbuf_copy(ptr, str, (int) strlen(str) + 1);
}

void strbuf_add_str(char **ptr, char *str) {
    strbuf_add(ptr, str, (int) strlen(str) + 1, 1);
}

void *strbuf_get_option(char *ptr) {
    strbuf_t *st = (strbuf_t *) (ptr - sizeof(strbuf_info_t));
    return st->inf.opt;
}

void strbuf_set_option(char *ptr, void *option) {
    strbuf_t *st = (strbuf_t *) (ptr - sizeof(strbuf_info_t));
    strbuf_info_t *inf;
    st->inf.opt = option;
    inf = GET_INTINF(st);
    inf->opt = option;
}
